src/utils/LimitTimer.hxx

```diff
@@ -0,0 +1,132 @@
+/** @copyright
+ * Copyright (c) 2020, Balazs Racz; 2021 Stuart Baker
+ * All rights reserved
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are  permitted provided that the following conditions are met:
+ *
+ *  - Redistributions of source code must retain the above copyright notice,
+ *    this list of conditions and the following disclaimer.
+ *
+ *  - Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * @file LimitTimer.hxx
+ *
+ * Limits the number of updates per unit time. Initial version written by
+ * Balazs Racz, modified for generalization by Stuart Baker
+ *
+ * @author Balazs Racz, Stuart Baker
+ * @date 9 January 2021
+ */
+
+#include "executor/Timer.hxx"
+ 
+/// This timer takes care of limiting the number of speed updates we send
+/// out in a second. It is a token bucket filter.
+class LimitTimer : public Timer
+{
+public:
+    /// Constructor.
+    /// @param ex executor to run on
+    /// @param update_delay_msec cooldown time delay in milliseconds
+    /// @param max_tokens number of available tokens
+    /// @param callback callback called once after cooldown time delay
+    LimitTimer(ExecutorBase *ex, uint16_t update_delay_msec, uint8_t max_tokens,
+               std::function<void()> callback)
+        : Timer(ex->active_timers())
+        , updateDelayMsec_(update_delay_msec)
+        , bucket_(max_tokens)
+        , bucketMax_(max_tokens)
+        , callback_(callback)
+    {
+        HASSERT(callback);
+    }
+
+    /// Destructor.
+    ~LimitTimer()
+    {
+        cancel();
+    }
+
+    /// Attempts to take a token out of the bucket.
```

Comment:
Please add a comment that this call must be made from the executor passed in on the constructor.

Reply:
Fixed.